File: 백준/Gold/3425. 고스택/고스택.cc
#include <bits/stdc++.h>
using namespace std;

const long long INF = 1000000000;
vector<string> cmds;
vector<long long> nums;

bool execute(long long start) {
    stack<long long> st;

    st.push(start);

    int idx = 0;
    for (string cmd : cmds) {
        if (cmd == "NUM") {
            st.push(nums[idx++]);
        } 
        else if (cmd == "POP") {
            if (st.empty()) return false;
            st.pop();
        } 
        else if (cmd == "INV") {
            if (st.empty()) return false;
            long long x = st.top(); st.pop();
            st.push(-x);
        } 
        else if (cmd == "DUP") {
            if (st.empty()) return false;
            st.push(st.top());
        } 
        else if (cmd == "SWP") {
            if (st.size() < 2) return false;
            long long a = st.top(); st.pop();
            long long b = st.top(); st.pop();
            st.push(a); st.push(b);
        } 
        else if (cmd == "ADD") {
            if (st.size() < 2) return false;
            long long a = st.top(); st.pop();
            long long b = st.top(); st.pop();
            long long res = a + b;
            if (abs(res) > INF) return false;
            st.push(res);
        } 
        else if (cmd == "SUB") {
            if (st.size() < 2) return false;
            long long a = st.top(); st.pop();
            long long b = st.top(); st.pop();
            long long res = b - a;
            if (abs(res) > INF) return false;
            st.push(res);
        } 
        else if (cmd == "MUL") {
            if (st.size() < 2) return false;
            long long a = st.top(); st.pop();
            long long b = st.top(); st.pop();
            long long res = a * b;
            if (abs(res) > INF) return false;
            st.push(res);
        } 
        else if (cmd == "DIV") {
            if (st.size() < 2) return false;
            long long a = st.top(); st.pop();
            long long b = st.top(); st.pop();
            if (a == 0) return false;
            long long res = abs(b) / abs(a);
            if ((b < 0) ^ (a < 0)) res = -res;
            if (abs(res) > INF) return false;
            st.push(res);
        } 
        else if (cmd == "MOD") {
            if (st.size() < 2) return false;
            long long a = st.top(); st.pop();
            long long b = st.top(); st.pop();
            if (a == 0) return false;
            long long res = abs(b) % abs(a);
            if (b < 0) res = -res;
            if (abs(res) > INF) return false;
            st.push(res);
        }
    }

    if (st.size() != 1) return false;

    cout << st.top() << "\n";
    return true;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    while (true) {
        cmds.clear();
        nums.clear();
        string cmd;
        while (cin >> cmd) {
            if (cmd == "QUIT") return 0;
            if (cmd == "END") break;
            cmds.push_back(cmd);
            if (cmd == "NUM") {
                long long x; cin >> x;
                nums.push_back(x);
            }
        }

        int n;
        cin >> n;
        for (int i = 0; i < n; i++) {
            long long x;
            cin >> x;
            if (!execute(x)) cout << "ERROR\n";
        }
        cout << "\n";
    }
    return 0;
}

File: 백준/Gold/3425. 고스택/고스택.java
import java.io.*;
import java.util.*;

// 고스택
//프로그램 에러
// 1. 숫자가 부족해서 연산 못함 -> stack의 갯수
// 2. 0으로 나눔(DIV, MOD)
// 3. 연산 결과의 절대값이 10^9 넘어갈때

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st;

        // 모든 기계
        boolean isEnd = false;
        while (!isEnd) {
            // 각 기계에서의 입력
            // 프로그램
            List<String> cmds = new ArrayList<>();
            while (true) {
                String input = br.readLine();

                if (input.equals("QUIT")) {
                    isEnd = true;
                    break;
                }

                if (input.equals("END")) {
                    break;
                }
                cmds.add(input);
            }

            if (isEnd) {
                break;
            }

            // 입력영역
            int n = Integer.parseInt(br.readLine());
            for (int i = 0; i < n; i++) {
                long num = Long.parseLong(br.readLine());

                Deque<Long> stack = new ArrayDeque<>();
                stack.add(num);

                boolean isError = false;
                for (String cmd : cmds) {
                    boolean error = false;

                    switch (cmd) {
                        case "POP":
                            if (stack.size() <= 0) {
                                error = true;
                            } else {
                                stack.removeFirst();
                            }
                            break;

                        case "INV":
                            if (stack.size() <= 0) {
                                error = true;
                            } else {
                                stack.addFirst(stack.removeFirst() * -1);
                            }
                            break;

                        case "DUP":
                            if (stack.size() <= 0) {
                                error = true;
                            } else {
                                stack.addFirst(stack.peekFirst());
                            }
                            break;

                        case "SWP":
                            if (stack.size() <= 1) {
                                error = true;
                            } else {
                                long a = stack.removeFirst();
                                long b = stack.removeFirst();
                                stack.addFirst(a);
                                stack.addFirst(b);
                            }
                            break;

                        case "ADD":
                            if (stack.size() <= 1) {
                                error = true;
                            } else {
                                long a = stack.removeFirst();
                                long b = stack.removeFirst();
                                stack.addFirst(a + b);
                            }
                            break;

                        case "SUB":
                            if (stack.size() <= 1) {
                                error = true;
                            } else {
                                long a = stack.removeFirst();
                                long b = stack.removeFirst();
                                stack.addFirst(b - a);
                            }
                            break;

                        case "MUL":
                            if (stack.size() <= 1) {
                                error = true;
                            } else {
                                long a = stack.removeFirst();
                                long b = stack.removeFirst();
                                stack.addFirst(a * b);
                            }
                            break;

                        case "DIV":
                            if (stack.size() <= 1 || stack.peekFirst() == 0) {
                                error = true;
                            } else {
                                long a = stack.removeFirst();
                                long b = stack.removeFirst();
                                long result = Math.abs(b) / Math.abs(a);
                                if (a * b < 0) {
                                    result *= -1;
                                }
                                stack.addFirst(result);
                            }
                            break;

                        case "MOD":
                            if (stack.size() <= 1 || stack.peekFirst() == 0) {
                                error = true;
                            } else {
                                long a = stack.removeFirst();
                                long b = stack.removeFirst();
                                long result = Math.abs(b) % Math.abs(a);
                                if (b < 0) {
                                    result *= -1;
                                }
                                stack.addFirst(result);
                            }
                            break;

                        default:    // NUM X 
                            st = new StringTokenizer(cmd);
                            st.nextToken(); // "NUM"
                            long newNum = Long.parseLong(st.nextToken());
                            stack.addFirst(newNum);
                            break;
                    }

                    // 연산결과 절대값 계산
                    if (stack.size() > 0 && Math.abs(stack.peekFirst()) > 1_000_000_000L) {
                        error = true;
                    }

                    if (error) {
                        isError = true;
                        break;
                    }
                }
                // 프로그램 실행 끝
                if (isError || stack.size() != 1) {
                    bw.write("ERROR\n");
                } else {
                    long a = stack.peekFirst();
                    bw.write(a + "\n");
                }

            }
            bw.write("\n");
            br.readLine();
        }

        bw.flush();
    }

}